File: exercises/max_area_of_island_695.ts
const DIRECTIONS: [number, number][] = [[-1, 0], [0, 1], [1, 0], [0, -1]];

function maxAreaOfIsland(grid: number[][]): number {
    if (!grid.length) return 0;

    const rows = grid.length;
    const cols = grid[0].length;
    let maxArea = 0;

    for (let y = 0; y < rows; y++) {
        for (let x = 0; x < cols; x++) {
            if (grid[y][x] == 0) {continue}
            maxArea = Math.max(maxArea, dfs(grid, y, x));
        }
    }

    return maxArea;
}
    
function dfs(grid: number[][], y: number, x: number): number {
    if (y < 0 || y >= grid.length || x < 0 || x >= grid[0].length) {
        return 0;
    }

    if (grid[y][x] === 0) return 0;

    grid[y][x] = 0;
    let area = 1;

    for (const [dy, dx] of DIRECTIONS) {
        area += dfs(grid, y + dy, x + dx);
    }

    return area;
}

// Iterative Solution really fast
//const DIRECTIONS = [-1, 0, 1, 0, -1];
//
//function maxAreaOfIsland(grid: number[][]): number {
//    const rows = grid.length;
//    const cols = grid[0].length;
//    let maxArea = 0;
//    let stack: number[] = new Array(rows * cols * 2);
//    let stackPtr = 0;
//
//    for (let y = 0; y < rows; y++) {
//        for (let x = 0; x < cols; x++) {
//            if (grid[y][x] === 0) continue;
//
//            let area = 0;
//            stackPtr = 0;
//            stack[stackPtr++] = y;
//            stack[stackPtr++] = x;
//
//            while (stackPtr > 0) {
//                const cx = stack[--stackPtr];
//                const cy = stack[--stackPtr];
//
//                if (cy < 0 || cy >= rows || cx < 0 || cx >= cols || grid[cy][cx] === 0) {
//                    continue;
//                }
//
//                grid[cy][cx] = 0;
//                area++;
//
//                for (let i = 0; i < 4; i++) {
//                    stack[stackPtr++] = cy + DIRECTIONS[i];
//                    stack[stackPtr++] = cx + DIRECTIONS[i + 1];
//                }
//            }
//
//            if (area > maxArea) maxArea = area;
//        }
//    }
//    return maxArea;
//}

File: exercises/max_area_of_island_695.java
// Recursive Solution
class Solution {
    private static final int[][] DIRECTIONS = { {-1, 0}, {0, 1}, {1, 0}, {0, -1} };
    
    public int maxAreaOfIsland(int[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }
        int maxArea = 0;
        int rows = grid.length;
        int cols = grid[0].length;
        for (int y = 0; y < rows; y++) {
            for (int x = 0; x < cols; x++) {
                maxArea = Math.max(maxArea, dfs(grid, y, x));
            }
        }
        return maxArea;
    }
    
    private int dfs(int[][] grid, int y, int x) {
        if (y < 0 || y >= grid.length || x < 0 || x >= grid[0].length) {
            return 0;
        }
        if (grid[y][x] == 0) {
            return 0;
        }
        grid[y][x] = 0;
        int area = 1;
        for (int[] dir : DIRECTIONS) {
            area += dfs(grid, y + dir[0], x + dir[1]);
        }
        return area;
    }
}

// Fast Iterative Solution
//class Solution {
//    private static final int[] DIRECTIONS = {-1, 0, 1, 0, -1};
//
//    public int maxAreaOfIsland(int[][] grid) {
//        int rows = grid.length;
//        int cols = grid[0].length;
//        int maxArea = 0;
//        int[] stack = new int[rows * cols * 2];
//        int stackPtr = 0;
//
//        for (int y = 0; y < rows; y++) {
//            for (int x = 0; x < cols; x++) {
//                if (grid[y][x] == 0) continue;
//
//                int area = 0;
//                stackPtr = 0;
//                stack[stackPtr++] = y;
//                stack[stackPtr++] = x;
//
//                while (stackPtr > 0) {
//                    int cx = stack[--stackPtr];
//                    int cy = stack[--stackPtr];
//
//                    if (cy < 0 || cy >= rows || cx < 0 || cx >= cols || grid[cy][cx] == 0) {
//                        continue;
//                    }
//
//                    grid[cy][cx] = 0;
//                    area++;
//
//                    for (int i = 0; i < 4; i++) {
//                        stack[stackPtr++] = cy + DIRECTIONS[i];
//                        stack[stackPtr++] = cx + DIRECTIONS[i + 1];
//                    }
//                }
//
//                if (area > maxArea) maxArea = area;
//            }
//        }
//        return maxArea;
//    }
//}
